### PR TITLE
Provide syntactic sugar for working with ChildViewControllers

### PR DESCRIPTION
**Pull request by marcocapano**
This requests adds two methods to quickly add and remove childviewcontrollers.
## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 4.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.
🚀 🚀 🚀

### CODE REVIEW

File: Sources/Extensions/UIKit/UIViewControllerExtensions.swift [hunk 1]
```diff
@@ -77,7 +77,27 @@ public extension UIViewController {
...
+    ///
+    /// - Parameters:
+    ///   - child: the view controller to add as a child
+    public func add(_ child: UIViewController) {
+        addChildViewController(child)
+        view.addSubview(child.view)
+        child.didMove(toParentViewController: self)
+    }
+    
+    /// SwifterSwift: Helper method to remove a UIViewController from its parent.
```
Comment (LucianoPAlmeida) [inline Sources/Extensions/UIKit/UIViewControllerExtensions.swift]: Here too, remove do not make clear that is from the parent and perform this operation, I would suggest removeFromParentViewController, but it already the method of UIKit. Maybe fullyRemoveFromParentViewController().

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift [hunk 2]
```diff
@@ -77,6 +77,38 @@ final class UIViewControllerExtensionsTests: XCTestCase {
...
+        
+        parentViewController.add(childViewController)
+        
+        XCTAssert(parentViewController.childViewControllers == [childViewController])
+        XCTAssertNotNil(childViewController.parent)
+        XCTAssertEqual(childViewController.parent, parentViewController)
+    }
+    
+    func testRemoveChildViewController() {
+        let parentViewController = UIViewController()
```
Comment (LucianoPAlmeida) [inline Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: Can you add a test case that covers parent == nil ?

**Comment by LucianoPAlmeida** [comment]
Just one more thing, you can run swiftlint autocorrect to fix the issues pointed by swiftlint on danger :)

File: Sources/Extensions/UIKit/UIViewControllerExtensions.swift (hunk 1 above)
Comment (SD10) [reply Sources/Extensions/UIKit/UIViewControllerExtensions.swift]: This is tough, I really don't like the name `fullyRemoveFromParentViewController`

File: Sources/Extensions/UIKit/UIViewControllerExtensions.swift [hunk 3]
```diff
@@ -77,7 +77,27 @@ public extension UIViewController {
         present(alertController, animated: true, completion: nil)
         return alertController
     }
-
+    
+    /// SwifterSwift: Helper method to add a UIViewController as a childViewController.
+    ///
+    /// - Parameters:
+    ///   - child: the view controller to add as a child
+    public func add(_ child: UIViewController) {
```
[the remark this answers is not in the document]
Comment (SD10) [reply Sources/Extensions/UIKit/UIViewControllerExtensions.swift]: Yeah we definitely can't use the name `add(_ child: UIViewController)` not descriptive enough. I think these are good extensions but we need better names for them

File: Sources/Extensions/UIKit/UIViewControllerExtensions.swift [hunk 4]
```diff
@@ -77,7 +77,27 @@ public extension UIViewController {
...
+        addChildViewController(child)
+        view.addSubview(child.view)
+        child.didMove(toParentViewController: self)
+    }
+    
+    /// SwifterSwift: Helper method to remove a UIViewController from its parent.
+    public func remove() {
+        guard parent != nil else {
+            return
```
Comment (SD10) [inline Sources/Extensions/UIKit/UIViewControllerExtensions.swift]: Can you move the `return` statement to the line above?

File: Sources/Extensions/UIKit/UIViewControllerExtensions.swift (hunk 1 above)
Comment (LucianoPAlmeida) [reply Sources/Extensions/UIKit/UIViewControllerExtensions.swift]: That's true, it's not so good 😆 But this one is tough, all the naming to make it more clear I could think is too big 😭 
What about `removeFromParentViewControllerWithView` or `removeViewAndControllerFromParentViewController`?

**Comment by guykogus** [comment]
Regarding the naming, I have these methods in my own project named like this:

```Swift
extension UIViewController {
    func addChildViewController(_ childViewController: UIViewController, toContainerView: UIView) {
        addChildViewController(childViewController)
        let childView = childViewController.view!
        toContainerView.addSubview(childView)
        childView.matchSuperview()
        childViewController.didMove(toParentViewController: self)
    }

    func removeViewFromParentViewController() {
        willMove(toParentViewController: nil)
        view.removeFromSuperview()
        removeFromParentViewController()
    }
}
```

**Comment by marcocapano** [comment]
I understand the naming issue. This one is pretty tough. I'm not sure about removeViewFromParentViewController suggested by @guykogus because it doesn't fell to me like it's saying that it removes the controller as childviewcontroller. That's why I would go for removeViewAndControllerFromParentViewController suggested by @LucianoPAlmeida . It is pretty long but at least is clear.
Same thing for the add method. I think addChildViewControllerWithView suggested by @LucianoPAlmeida expresses the fact that the method apart from the addChildViewController also calls addSubview and didMoveToSuperview on the childViewController.

File: Sources/Extensions/UIKit/UIViewControllerExtensions.swift (hunk 4 above)
Comment (marcocapano) [reply Sources/Extensions/UIKit/UIViewControllerExtensions.swift]: sure!

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift (hunk 2 above)
Comment (marcocapano) [reply Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: I added a test case that checks if everything stays the same if you call remove on a view controller that has no parent. Is that what you meant?

**Comment by marcocapano** [comment]
maybe a containerView parameter with a default value of self.view?

**Comment by marcocapano** [comment]
Do you think the presence of not of this parameter should affect the naming?

**Comment by marcocapano** [comment]
I feel like two versions of the same function in this case would not be good. The naming is already difficult like this. I would keep things simple.

Would you like going for a slightly modified version on @guykogus naming?

```
public func addChildViewController(_ child: UIViewController, toContainerView containerView: UIView) {
    addChildViewController(child)
    containerView.addSubview(child.view)
    child.didMove(toParentViewController: self)
}
```

I added the second parameter name because I don't like using toContainerView as a name in the function body.

**Comment by guykogus** [comment]
I agree on the parameter naming.

By the way, GitHub can format your code blocks in the comments for a given language. Check out [this doc](https://help.github.com/articles/creating-and-highlighting-code-blocks/) ;)

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift [hunk 5]
```diff
@@ -77,6 +77,47 @@ final class UIViewControllerExtensionsTests: XCTestCase {
...
+    func testAddChildViewController() {
+        let parentViewController = UIViewController()
+        let childViewController = UIViewController()
+
+        XCTAssert(parentViewController.childViewControllers.isEmpty == true)
+        XCTAssertNil(childViewController.parent)
+        
+        parentViewController.addChildViewController(childViewController, toContainerView: parentViewController.view)
+        
+        XCTAssert(parentViewController.childViewControllers == [childViewController])
```
Comment (guykogus) [inline Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: Use `XCTAssertEqual` here instead.

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift [hunk 6]
```diff
@@ -77,6 +77,47 @@ final class UIViewControllerExtensionsTests: XCTestCase {
...
+        XCTAssertNotNil(childViewController.parent)
+        XCTAssertEqual(childViewController.parent, parentViewController)
+    }
+    
+    func testRemoveChildViewController() {
+        let parentViewController = UIViewController()
+        let childViewController = UIViewController()
+        
+        parentViewController.addChildViewController(childViewController)
```
Comment (guykogus) [inline Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: You don't want to use the new `addChildViewController` method that you created?

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift [hunk 7]
```diff
@@ -77,6 +77,47 @@ final class UIViewControllerExtensionsTests: XCTestCase {
...
+        parentViewController.view.addSubview(childViewController.view)
+        childViewController.didMove(toParentViewController: parentViewController)
+        
+        XCTAssert(parentViewController.childViewControllers == [childViewController])
```
Comment (guykogus) [inline Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: If you use the new `addChildViewController` method, which you already tested in another function, you don't need these checks, technically. If you're going to use them all together you could even combine the 2 test functions into this 1, where you test both at the same time.

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift [hunk 8]
```diff
@@ -77,6 +77,47 @@ final class UIViewControllerExtensionsTests: XCTestCase {
...
+        XCTAssertNotNil(childViewController.parent)
+        XCTAssertEqual(childViewController.parent, parentViewController)
+        
+        childViewController.removeViewAndControllerFromParentViewController()
+        
+        XCTAssert(parentViewController.childViewControllers.isEmpty == true)
```
Comment (guykogus) [inline Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: No need for `== true`.

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift [hunk 9]
```diff
@@ -77,6 +77,47 @@ final class UIViewControllerExtensionsTests: XCTestCase {
 		}
 		XCTAssertEqual(alertController.preferredAction, alertController.actions[preferredButtonIndex])
 	}
+    
+    func testAddChildViewController() {
+        let parentViewController = UIViewController()
+        let childViewController = UIViewController()
+
+        XCTAssert(parentViewController.childViewControllers.isEmpty == true)
```
Comment (guykogus) [inline Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: No need for `== true`

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift (hunk 6 above)
Comment (marcocapano) [reply Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: To be more clear: We do this with the implementation(we don't use a method in another method of the library) so why not with tests. This way we can safely remove one later.

File: CHANGELOG.md [hunk 10]
```diff
@@ -11,6 +11,9 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
 - **UIView**
   - Added `addGestureRecognizers(_:)` which accepts an array of `UIGestureRecognizer` to add multiple gesture recognizers to a view with one call. [#523](https://github.com/SwifterSwift/SwifterSwift/pull/523) by [moyerr](https://github.com/moyerr)
   - Added `removeGestureRecognizers(_:)` which accepts an array of `UIGestureRecognizer` to remove multiple gesture recognizers from a view with one call. [#523](https://github.com/SwifterSwift/SwifterSwift/pull/523) by [moyerr](https://github.com/moyerr)
+- **UIViewController**
+  - Added addChildViewController(_ child: , toContainerView:) which accepts a UIViewController, to easily add child view controllers.
```
Comment (guykogus) [inline CHANGELOG.md]: Please use back-ticks (`) for function and class names, like in the other comments.

File: Tests/UIKitTests/UIViewControllerExtensionsTests.swift (hunk 6 above)
Comment (guykogus) [reply Tests/UIKitTests/UIViewControllerExtensionsTests.swift]: I guess that's a fair point. It's OK like this, then.

File: CHANGELOG.md (hunk 10 above)
Comment (marcocapano) [reply CHANGELOG.md]: Sorry for that. This is my first time writing a Changelog 😅

File: CHANGELOG.md [hunk 11]
```diff
@@ -11,6 +11,9 @@ The changelog for **SwifterSwift**. Also see the [releases](https://github.com/S
 - **UIView**
   - Added `addGestureRecognizers(_:)` which accepts an array of `UIGestureRecognizer` to add multiple gesture recognizers to a view with one call. [#523](https://github.com/SwifterSwift/SwifterSwift/pull/523) by [moyerr](https://github.com/moyerr)
   - Added `removeGestureRecognizers(_:)` which accepts an array of `UIGestureRecognizer` to remove multiple gesture recognizers from a view with one call. [#523](https://github.com/SwifterSwift/SwifterSwift/pull/523) by [moyerr](https://github.com/moyerr)
+- **UIViewController**
+  - Added `addChildViewController(_ child: , toContainerView:)` which accepts a UIViewController, to easily add child view controllers.
```
Comment (guykogus) [inline CHANGELOG.md]: Sorry for being pedantic here. It should be `addChildViewController(_:toContainerView)` and with backticks for `UIViewController` as well (in the line below too).

Comment (marcocapano) [reply CHANGELOG.md]: I'm here to learn, feel free to destroy my pull request😂😂

File: Sources/Extensions/UIKit/UIViewControllerExtensions.swift [hunk 12]
```diff
@@ -78,6 +78,24 @@ public extension UIViewController {
         return alertController
     }
 
+    /// SwifterSwift: Helper method to add a UIViewController as a childViewController.
+    ///
+    /// - Parameters:
+    ///   - child: the view controller to add as a child
```
Comment (LucianoPAlmeida) [inline Sources/Extensions/UIKit/UIViewControllerExtensions.swift]: Can you add the container view parameter in the docs :)

**Comment by LucianoPAlmeida** [comment]
@marcocapano I see you also putting other extensions on the same PR. But acctually we prefer to have diferent extensions in diferent PR's as we mention "one" extension per PR in the [Contributing Guidelines](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md#adding-new-extensions) 👍 
Feel free to open another with this UIDevice extension :))

**Comment by marcocapano** [comment]
@LucianoPAlmeida I was actually working on a different branch but accidentally committed on this one. I already removed it 👍 Also included containerView in the documentation

**Comment by guykogus** [comment]
I think he meant adding reference to the function accepting a container view. Can you add a mention to that?

**Comment by marcocapano** [comment]
>I think he meant adding reference to the function accepting a container view. Can you add a mention to that?

@guykogus of course! But not before this evening 🙏🏻

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by marcocapano** [comment]
@guykogus done 👍🏻

**Comment by marcocapano** [comment]
Should I ask someone to merge? I don’t think I have permissions, right?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by SD10** [comment]
@marcocapano As of now, we only allow core team members to merge to `master` after there have been 2+ approving reviews
